### PR TITLE
Fix HiCache storage invalidation after weight update

### LINKED ISSUE #26792: [Bug] HiCache persistent storage can reload stale KV pages after online weight update
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

When SGLang runs with HiCache persistent file storage enabled, `update_weights_from_tensor` with `flush_cache=True` clears the live GPU/host KV cache but does not invalidate the persistent HiCache file storage. A later request with the same prompt can reload KV pages that were computed under the previous model weights.

I confirmed this on current main commit `7ed53d15f357ea4d722c1980c2cb35e8367d8bb0` with `Qwen/Qwen3-30B-A3B` on a single NVIDIA A100 80GB.

Observed evidence:

- Request A used the original weights and wrote persistent HiCache pages.
- The online update succeeded with `flush_cache=True`, `abort_all_requests=True`, and `weight_version="step-1"`.
- The updated tensor was `model.layers.0.self_attn.qkv_proj.weight`, shape `[5120, 2048]`, set to CUDA bf16 zeros.
- HiCache file storage had 4166 files after A.
- HiCache file storage still had 4166 files after the weight update.
- 4166 old files survived the update; 0 new files were created by the update itself.
- Request B used the same prompt after the update and reported 4157 cached tokens from persistent storage:

```json
{
  "device": 0,
  "host": 0,
  "storage": 4157,
  "storage_backend": "HiCacheFile"
}
```

Controls after the same weight update did not use storage and agreed with each other:

- C: clear storage directory before the post-update request, cached tokens 0.
- D: disable HiCache storage backend before the post-update request, cached tokens 0.
- E: fresh server with updated weights and empty storage, cached tokens 0.

Checksums over output token IDs plus output logprobs/top-logprobs:

- B, storage intact after update: `aaaafc985422c90b9ce6f639d98fc55d19841b899cf9c5036481d1cd751aba70`
- C, storage cleared: `220c9966a442040d27d0a21802de2082fbe6973f775eb98c2178c35ef7db630b`
- D, storage disabled: `220c9966a442040d27d0a21802de2082fbe6973f775eb98c2178c35ef7db630b`
- E, fresh updated server: `220c9966a442040d27d0a21802de2082fbe6973f775eb98c2178c35ef7db630b`

The generated output token IDs happened to remain the same in this run, but logprobs/top-logprobs diverged. This still exposes a correctness issue: request B reused old-weight prefix KV and produced logits/logprobs different from all updated-weight controls.

Additional instrumentation showed that the HiCache file key suffix was model/parallelism based, not `weight_version` based:

```text
_qwen3-30b-a3b-hicache-debug_0_1
```

The same first storage keys appeared before and after the weight update. The read path reported storage files whose mtimes predated the update timestamp.

Static code audit from the same commit:

- `flush_cache_after_weight_update()` delegates to scheduler `flush_cache()` when `flush_cache=True`.
- Scheduler `flush_cache()` resets live radix/pool state and metrics, but does not clear the HiCache persistent file backend.
- HiCache file names include token/page hash plus model/TP/PP suffix, but no observed `weight_version` component.
- Persistent storage clearing exists as a separate admin path, not as part of weight-update flush.

Expected behavior:

After model weights change, KV cache computed under old weights must not be reused. `flush_cache=True` after an online weight update should invalidate every cache level that can affect generation under the new weights, including persistent HiCache storage. Reasonable fixes would be to clear persistent storage during weight-update flush, include `weight_version` or model identity in persistent storage keys, or reject/disable persistent HiCache storage across online weight updates until it is explicitly reset.

### Reproduction

I used a local validation harness equivalent to the following flow.

1. Clear the HiCache file storage directory.
2. Start SGLang with hierarchical cache and file storage:

```bash
export SGLANG_HICACHE_FILE_BACKEND_STORAGE_DIR=/home/gabe/hicache-test

python -m sglang.launch_server \
  --model-path Qwen/Qwen3-30B-A3B \
  --host 127.0.0.1 \
  --port 30000 \
  --dtype bfloat16 \
  --trust-remote-code \
  --served-model-name qwen3-30b-a3b-hicache-debug \
  --admin-api-key hicache-debug-admin \
  --enable-hierarchical-cache \
  --hicache-storage-backend file \
  --hicache-write-policy write_through \
  --hicache-storage-prefetch-policy wait_complete \
  --hicache-size 80 \
  --mem-fraction-static 0.88 \
  --context-length 8192 \
  --max-running-requests 1 \
  --disable-cuda-graph \
  --enable-deterministic-inference
```

3. Send request A with a deterministic long prompt, 4158 prompt tokens, `temperature=0`, and small `max_new_tokens`.
4. Wait until HiCache file storage is stable.
5. Apply an online weight update:

```python
from sglang.srt.utils import MultiprocessingSerializer
import requests
import torch

name = "model.layers.0.self_attn.qkv_proj.weight"
tensor = torch.zeros((5120, 2048), dtype=torch.bfloat16, device="cuda")
payload = {
    "serialized_named_tensors": [
        MultiprocessingSerializer.serialize([(name, tensor)], output_str=True)
    ],
    "load_format": "direct",
    "flush_cache": True,
    "abort_all_requests": True,
    "weight_version": "step-1",
    "torch_empty_cache": False,
}
requests.post(
    "http://127.0.0.1:30000/update_weights_from_tensor",
    json=payload,
    headers={"Authorization": "Bearer hicache-debug-admin"},
    timeout=600,
)
```

6. Send request B with the same prompt and sampling while keeping storage intact.
7. Compare with controls:
   - C: same update, but clear storage before the post-update request.
   - D: same update, but delete/disable the HiCache storage backend before the post-update request.
   - E: fresh updated server with empty storage.

The confirmation criteria were:

- B gets persistent storage hits for pages written before the update.
- Storage keys are identical before and after the update.
- B differs from the clear-storage/no-storage/fresh-updated controls.

All three criteria were met.

### Environment

The raw `python3 -m sglang.check_env` output was not retained before the GCP VM was deleted, but the run environment was:

- SGLang: editable checkout of `sgl-project/sglang` main at `7ed53d15f357ea4d722c1980c2cb35e8367d8bb0`
- Hardware: GCP `a2-ultragpu-1g`, single NVIDIA A100-SXM4-80GB
- Model: `Qwen/Qwen3-30B-A3B`
- Python: 3.10 paths in logs (`/home/gabe/.local/lib/python3.10/site-packages/...`)
- PyTorch/CUDA: CUDA-enabled PyTorch; server log included a warning that PyTorch was built with CUDA 13.0 while TorchAudio was built with CUDA 12.9
- Attention backend selected by server: prefill `fa3`, decode `flashinfer`
- Dtype: `bfloat16`
- Context length: 8192
- HiCache settings: hierarchical cache enabled, file storage backend, `write_through`, `wait_complete`, `hicache_size=80`

I searched open and closed issues/PRs for:

- `HiCache update_weights`
- `HiCache weight_version`
- `persistent storage weight update`
- `stale KV update_weights_from_tensor`
- `flush_cache_after_weight_update storage`
- `update_weights_from_disk HiCache`

I did not find an existing matching report.

### PR DESCRIPTION
**Pull request by SouthWest7**
## Summary

- Clear HiCache persistent storage as part of online weight-update cache flushing.
- Propagate explicit storage-backend clear failures so stale persistent KV is not silently reused after weights change.
- Make `HiCacheFile.clear()` tolerant of concurrent file deletion during multi-rank clears.

Fixes sgl-project/sglang#26792.

## Root Cause

`update_weights_from_tensor(..., flush_cache=True)` clears the live GPU/host KV cache through scheduler `flush_cache()`, but persistent HiCache storage remains addressable with the same token/page keys. A later request with the same prompt can therefore reload KV pages computed under old model weights.

## Validation

- `python -m py_compile` on the touched Python files.
- `git diff --check upstream/main...HEAD` and `git diff --check`.
- Manual E2E on Qwen/Qwen3-30B-A3B with HiCache file storage, after merging latest upstream `main`:
  - Request A populated storage: `4173` files / `410222592` bytes.
  - `update_weights_from_tensor` with `flush_cache=True`, `abort_all_requests=True`, and `weight_version='step-1'` returned success.
  - Storage after update: `0` files / `0` bytes.
  - Request B with the same prompt reported `cached_tokens: 0` and `cached_tokens_details: None`.
  - Negative control with `flush_cache=False` preserved storage: `4159` files / `408846336` bytes before and after update.
  - Final E2E output: `E2E_ALL_PASS`.

---
### CI States

Latest PR Test (Base): :x: [Run #28254382218](https://github.com/sgl-project/sglang/actions/runs/28254382218)
Latest PR Test (Extra): :x: [Run #28254382122](https://github.com/sgl-project/sglang/actions/runs/28254382122)

### CODE REVIEW

**Comment by ronhuafeng** [comment]
Validated PR head `33b7372d19dfc77d4f05c987012cd2333adb322b` locally as a lightweight supplement to the E2E validation already recorded in the PR body.

Checks I ran:
- `git diff --check c98d31143..HEAD`
- `py_compile` on the changed Python files
- changed-file `pre-commit run --files ...`
- `pytest test/registered/unit/managers/test_scheduler_flush_cache.py -q` -> 7 passed
- ad-hoc invariant checks for `SchedulerWeightUpdaterManager.flush_cache_after_weight_update()`:
  - `flush_cache=False` does not clear live cache or HiCache storage
  - `flush_cache=True` calls live cache flush and then HiCache storage clear
  - live-cache failure and storage-clear failure both raise instead of silently continuing
  - disabled hierarchical cache / disabled HiCache storage are no-op success cases
  - storage-enabled path propagates `tree_cache.clear_storage_backend()` success/failure
- ad-hoc `HiCacheFile.clear()` race check: a concurrent `FileNotFoundError` during unlink is tolerated and evictor metadata is still cleared

Static read: all online weight-update sources share `flush_cache_after_weight_update()`, so the persistent-storage invalidation covers disk, distributed, tensor, and IPC update paths when `flush_cache=True`. I did not rerun the full model E2E; the PR body already records that end-to-end run. GitHub lint is green, and the remaining red checks appear to be `run-ci` gate / skipped-job aggregate fallout rather than executed test failures.
